mqtt/mqtt-bridge/src/persist/memory/waking_map.rs

```diff
@@ -0,0 +1,225 @@
+use std::cmp::min;
+use std::collections::VecDeque;
+use std::{collections::HashMap, task::Waker};
+
+use mqtt3::proto::Publication;
+use thiserror::Error;
+
+use crate::persist::Key;
+
+// Responsible for waking waiting streams when new elements are added
+pub struct WakingMap {
+    queue: VecDeque<(Key, Publication)>,
+    in_flight: HashMap<Key, Publication>,
+    waker: Option<Waker>,
+}
+
+impl WakingMap {
+    pub fn new() -> Self {
+        let queue: VecDeque<(Key, Publication)> = VecDeque::new();
+        let in_flight = HashMap::new();
+
+        WakingMap {
+            queue,
+            in_flight,
+            waker: None,
+        }
+    }
+
+    pub fn insert(&mut self, key: Key, value: Publication) {
+        self.queue.push_back((key, value));
+
+        if let Some(waker) = self.waker.take() {
+            waker.wake();
+        }
+    }
+
+    pub fn get(&mut self, count: usize) -> Result<Vec<(Key, Publication)>, WakingMapError> {
+        let count = min(count, self.queue.len());
+        let mut output = vec![];
+        for _ in 0..count {
+            let removed = self
+                .queue
+                .pop_front()
+                .ok_or(WakingMapError::InFlightConversion())?;
+            output.push(removed);
+        }
+
+        for element in output.iter() {
+            self.in_flight.insert(element.0.clone(), element.1.clone());
+        }
+
+        Ok(output)
+    }
+
+    pub fn remove_in_flight(&mut self, key: &Key) -> Result<(), WakingMapError> {
+        self.in_flight
+            .remove(key)
+            .ok_or(WakingMapError::InFlightRemoval())?;
+        Ok(())
+    }
+
+    pub fn set_waker(&mut self, waker: &Waker) {
+        self.waker = Some(waker.clone());
+    }
+
+    // TODO REVIEW: exposed just for testing.
+    // We don't use the in-flight now, but we need to find a way to use it.
+    // Otherwise just have the queue.
+    #[allow(dead_code)]
+    fn in_flight(&mut self) -> &HashMap<Key, Publication> {
+        &self.in_flight
+    }
```

Comment:
The in-flight messages for the in-memory persistence implementation don't have a clear purpose. 

For the disk case (not this PR) it makes sense. For disk case, messages cannot be removed from the main collection right away because in case of crash we want them to be resent. Thus a separate in-flight collection is needed to prevent duplicates from being returned from the loader. When an element is removed form the in-flight it can be removed from the main persistence.

For the in-mem case everything is lost in the case of a crash anyway, so the in-flight collection doesn't have the same purpose. Removing from the queue immediately on retrieval by the loader allows for a simple queue to be used. The messages can be removed from the main queue directly when retrieved by the loader, rather than linger around in the main queue until being removed from in-flight. I think having this collection is still necessary for logging, but we need to find an appropriate place to do that. 

If we cannot think of a place to log in-flight, then we can remove the in-flight collection from this in-memory implementation.